### ISSUE TITLE
Fetch app updates prior to starting Nextcloud upgrade

### ISSUE THREAD

**Issue by blizzz** [nextcloud/updater#580]
To combat issues like https://github.com/nextcloud/server/issues/47602 we should also fetch necessary app archives around downloading the nextcloud release.

A brainstorming with @nickvergessen  brought  up these ideas:

- Have a step to download app archives around download the nextcloud server archive
- To do so it is necessary to query the running instance to get the enabled apps
- ~~Allow to specify apps that shall be ignored per parameter~~ Report, but otherwise ignore apps that are not available on appstore
- When appstore is unavailable or an app download failed report and:
  -  ask whether to continue in interactive mode
  -  In non-interactive mode fail, 
  -  continue when `--force` is specified.
- app folder need to be read from config.php, if something is specified then i the first writable location shall be used.

**Comment by blizzz** [comment]
cc @Altahrim

**Comment by joshtrichards** [comment]
Related: #401 & #456

> To do so it is necessary to query the running instance to get the enabled apps

The Updater itself technically doesn't handle upgrading of the apps, have db access, nor have access to any of our usual APIs. We should probably handle all this prior to even invoking Updater let alone `occ upgrade`.

We do have a convenient spot before Updater: the `updatenotification` app, which already checks apps that need updating and compatibility. We could also pre-download apps there (or follow the same approach anyway, by handling all this before the Updater is even invoked).

> app folder need to be read from config.php, if something is specified then i the first writable location shall be used. Placing them there is only possible after replacing code. Easier (for now?) is to add them just to the apps directory

Not  quite following the context for this one. Are you talking about where to drop the Archive files for the new pre-downloaded apps?

**Comment by blizzz** [comment]
> > To do so it is necessary to query the running instance to get the enabled apps
> 
> The Updater itself technically doesn't handle upgrading of the apps, have db access, nor have access to any of our usual APIs. We should probably handle all this prior to even invoking Updater let alone `occ upgrade`.

Yes, that is true, but it is already download the server code, so downloading relevant apps is not so far off either. 

The issue with loading them before invoking the updater has the issue that the apps have to be put somewhere, and not replace current apps (they might not be compatible). There would be ways around this too, though.

> We do have a convenient spot before Updater: the `updatenotification` app, which already checks apps that need updating and compatibility. We could also pre-download apps there (or follow the same approach anyway, by handling all this before the Updater is even invoked).

For the `updatenotification`  can also be disabled, so we cannot exclusively rely on it.

> > app folder need to be read from config.php, if something is specified then i the first writable location shall be used. Placing them there is only possible after replacing code. Easier (for now?) is to add them just to the apps directory
> 
> Not quite following the context for this one. Are you talking about where to drop the Archive files for the new pre-downloaded apps?

Yes, they have to be put in place after replacing the nextcloud server code.

**Comment by joshtrichards** [comment]
Related: nextcloud/server#33561

**Comment by joshtrichards** [comment]
I think we should try to do as much as possible before the code update (and, from a user experience perspective, as much as we can before even offering an update). 

Also, probably good opportunity to move some/all of the downloading into a "non-screen time" window.

Ideally, we make the solution(s) compatible with the upgrade mechanisms used by all the packaging/deploy methods (i.e. Docker, AIO, Snap, etc.). I say this because the **code updates of the apps at Server upgrade time is the one thing that *all* of the deployment methods have in common.** 

So I'm thinking we come up with a solution that permits:

- checking for compatible app availability (ahead of code update)
- pre-downloading apps (prior to code update) 
- minimizing likelihood of a failed app update resulting in outright data loss (when the new version is installed)

The hardest problem to solve is probably the last one. We could probably do the first two items in a "v1.0" and leave the last one for "v2.0".

There's also probably an opportunity here to start pre-downloading Server archive too (i.e. in the same background job we implement for the apps - albeit with a toggle or something for the deploy methods that only need it for the apps).
